runtime/collections/rangemaps.cpp

```diff
@@ -0,0 +1,282 @@
+  block *end;
+};
+static struct blockheader range_header() {
+  static struct blockheader hdr = {(uint64_t)-1};
+  if (hdr.hdr == -1) {
+    hdr = getBlockHeaderForSymbol(
+        (uint64_t)getTagForSymbolName("LblRangemap'Coln'Range{}"));
+  }
+  return hdr;
+}
+rangemap hook_RANGEMAP_elementRng(SortKItem rng, SortKItem value) {
+  range *ptr = (range *)rng;
+  return hook_RANGEMAP_element(ptr->start, ptr->end, value);
+}
+
+rangemap hook_RANGEMAP_unit() {
+  return rangemap();
+}
+
+rangemap hook_RANGEMAP_concat(SortRangeMap m1, SortRangeMap m2) {
+  auto mfirst = m1->size() >= m2->size() ? m1 : m2;
+  auto msec = m1->size() >= m2->size() ? m2 : m1;
+  return mfirst->concat(*msec);
+}
+
+SortKItem hook_RANGEMAP_lookup_null(SortRangeMap m, SortKItem key) {
+  auto val = m->get_value(key);
+  if (val.has_value()) {
+    return val.value();
+  }
+  return nullptr;
+}
+
+SortKItem hook_RANGEMAP_lookup(SortRangeMap m, SortKItem key) {
+  auto res = hook_RANGEMAP_lookup_null(m, key);
+  if (!res) {
+    KLLVM_HOOK_INVALID_ARGUMENT("Key not found for range map lookup");
+  }
+  return res;
+}
+
+SortKItem hook_RANGEMAP_lookupOrDefault(
+    SortRangeMap m, SortKItem key, SortKItem _default) {
+  auto res = hook_RANGEMAP_lookup_null(m, key);
+  if (!res) {
+    return _default;
+  }
+  return res;
+}
+
+SortRange hook_RANGEMAP_find_range(SortRangeMap m, SortKItem key) {
+  auto val = m->get_key_value(key);
+  if (val.has_value()) {
+    range *ptr = (range *)koreAlloc(sizeof(range));
+    ptr->h = range_header();
+    ptr->start = val.value().first.start();
+    ptr->end = val.value().first.end();
+    return (SortRange)ptr;
+  } else {
+    KLLVM_HOOK_INVALID_ARGUMENT("Key not found for range map lookup");
+  }
+}
+
+rangemap hook_RANGEMAP_update(
+    SortRangeMap m, SortKItem start, SortKItem end, SortKItem value) {
+  return m->inserted(rng_map::Range<KElem>(start, end), value);
+}
+
+rangemap
+hook_RANGEMAP_updateRng(SortRangeMap m, SortKItem rng, SortKItem value) {
+  range *ptr = (range *)rng;
+  return hook_RANGEMAP_update(m, ptr->start, ptr->end, value);
+}
+
+rangemap hook_RANGEMAP_remove(SortRangeMap m, SortKItem start, SortKItem end) {
+  return m->deleted(rng_map::Range<KElem>(start, end));
+}
+
+rangemap hook_RANGEMAP_removeRng(SortRangeMap m, SortKItem rng) {
+  range *ptr = (range *)rng;
+  return hook_RANGEMAP_remove(m, ptr->start, ptr->end);
+}
+
+rangemap hook_RANGEMAP_difference(SortRangeMap m1, SortRangeMap m2) {
+  return m1->difference(*m2);
+}
+
+set hook_SET_unit(void);
+set hook_SET_element(block *);
+set hook_SET_concat(set *, set *);
+
+set hook_RANGEMAP_keys(SortRangeMap m) {
+  auto tmp = hook_SET_unit();
+  for (auto iter = rng_map::ConstRangeMapIterator<KElem, KElem>(*m);
+       iter.has_next(); ++iter) {
+    range *ptr = (range *)koreAlloc(sizeof(range));
+    ptr->h = range_header();
+    ptr->start = iter->first.start();
+    ptr->end = iter->first.end();
+    auto elem = hook_SET_element((SortRange)ptr);
```

Comment:
this is also not quite right. a `Set` contains elements of sort `KItem`, but here you are putting elements of sort Range into it. You need to create an injection from Range into KItem to wrap around it.

Reply:
Addressed, adding the missing injection to the set items.